[Merged by Bors] - Support poet certificates

poszu opened the pull request:
## Motivation
Closes #5204 

## Changes
Added support for registering in the poet services using a certificate instead of PoW. It still supports both ways and will fallback to the PoW if:
- certificate is not available (couldn't certify)
- the poet service doesn't support certificates

There is a new interface in `activation` package `certifierService` that provides a generic way to obtain a certificate for a given poet. It abstracts away the fact that a POST proof is used - it is hidden in `certifierServiceClient`, which could in future have different implementations for other poet services.

Changed the `initial_post` table to `post` as it now holds the latest POST proof. Relying solely on initial post proof OR an existing ATX could not work in case of a checkpoint (in fact this way it failed systests).

## Test Plan
- added UTs
- added system tests

poszu committed:
Support for poet certificates WIP

poszu committed:
Support for poet certificates cont

poszu commented:
bors try

poszu committed:
Initial certification using existing ATX or initial POST

poszu commented:
bors try

poszu committed:
Improve poet HTTP client coverage

poszu committed:
Bump certifier

poszu committed:
Use the right post challenge to certify

poszu committed:
Fixed POST metadata sent to certifier

poszu committed:
Bump certifier image

poszu committed:
Migrate certifier store to localdb

poszu committed:
Remove CertifierInfo struct

poszu committed:
Add systest for poet registrations with PoW and cert

poszu commented:
bors try

poszu commented:
bors try

poszu committed:
Configurable certifier retry options

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu commented:
bors try

poszu committed:
Increase systest CI job timeout

poszu commented:
bors try

poszu committed:
Keep POST proof in localDB for certification

poszu committed:
Take optional poet certificates in config

poszu commented:
bors try

poszu committed:
Bump poet to v0.10.0

fasmat commented on sql/migrations/local/0002_next.sql:
```diff
@@ -7,3 +7,15 @@ ALTER TABLE initial_post RENAME COLUMN commit_atx_new TO commit_atx;
 ALTER TABLE initial_post ADD COLUMN vrf_nonce        UNSIGNED LONG INT NOT NULL;
+ALTER TABLE initial_post ADD COLUMN challenge        VARCHAR NOT NULL;
+
+ALTER TABLE initial_post RENAME TO post;
```
This collides with https://github.com/spacemeshos/go-spacemesh/pull/5266.

You will probably have to add a 0003 migration and do the changes there.

poszu replied:
Moved to 0003_next.sql.

fasmat commented on sql/localsql/nipost/nipost.go:
```diff
@@ -8,71 +8,65 @@ import (
-		insert into initial_post (
-			id, post_nonce, post_indices, post_pow, num_units, commit_atx, vrf_nonce
-		) values (?1, ?2, ?3, ?4, ?5, ?6, ?7);`, enc, nil,
+		REPLACE into post (
+			id, post_nonce, post_indices, post_pow, challenge, num_units, commit_atx, vrf_nonce
+		) values (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8);`, enc, nil,
```
I'm not sure about using `REPLACE` instead of `INSERT`. I think this makes it too easy to overwrite an existing challenge by accident.

poszu replied:
Wdym? The point is to overwrite the existing POST proof with a new one.

fasmat replied:
In my opinion this should never be necessary if we just keep what is in `initial_post`:

- A new node will create an initial post that can be kept forever in localDB to be able to create certification requests
- An existing node will import their initial post during startup with a database migration and it will also be kept forever

Even with nodes that have to resync this is quite easy:

1. During the "create initial post" step if no post is in `initial_post`, check if this is the first ATX for an identity
2. If not use any previous post of the identity

This only breaks in one specific case: The node just restored a checkpoint AND the operator decided to delete their `localsql` (which they shouldn't do after a checkpoint). In that specific case we can fall back to creating a new initial post only for the sake of having one for the purpose of requesting a certificate.

poszu replied:
It's not only this specific case. The user will not find the initial POST after checkpoint if:
- he decided to remove post.bin in the past (which is fine)
- he already ran v1.3 - this version removes the initial post from the DB by itself after publishing an ATX.

The second point is the real problem (the code removing the initial post from DB is already used in prod). 

IMO, there is no downside to keeping the last POST proof in the local DB and it can improve UX of these users (they will avoid regenerating the initial post after checkpoint).

fasmat commented on .github/workflows/systest.yml:
```diff
@@ -108,7 +108,7 @@ jobs:
 
       - name: Run tests
-        timeout-minutes: 60
+        timeout-minutes: 80
```
The additional tests add ~ 10 minutes to the execution time of the system tests. Up to ~ 55 min from ~45 min. Is it possible to run the new tests in parallel to the existing ones?

poszu replied:
I think that any systest could be run in parallel as each runs in its own namespace. I'm not sure what is the policy to decide.

fasmat commented on activation/activation.go:
```diff
@@ -343,7 +358,7 @@ func (b *Builder) buildInitialPost(ctx context.Context) error {
 	}
 	// ...and if we haven't stored an initial post yet.
-	_, err := nipost.InitialPost(b.localDB, b.signer.NodeID())
+	_, err := nipost.GetPost(b.localDB, b.signer.NodeID())
```
Getters in go usually drop the `Get-` prefix: https://go.dev/doc/effective_go#Getters

```suggestion
	_, err := nipost.Post(b.localDB, b.signer.NodeID())
```

poszu replied:
Sure, but the problem is that `Post` already exists in package `nipost`.

fasmat replied:
This is related to the other point: this should really just be the initial post of a node, and for most nodes it will be. Putting another post into this table is just a fallback for some nodes that otherwise would have to regenerate it.

fasmat commented on activation/activation.go:
```diff
@@ -365,15 +380,91 @@ func (b *Builder) buildInitialPost(ctx context.Context) error {
+func (b *Builder) certifyPost(ctx context.Context) {
+	post, meta, ch, err := b.obtainPostForCertification()
+	if err != nil {
+		b.log.With().Error("failed to obtain post for certification", zap.Error(err))
```
```suggestion
		b.log.Error("failed to obtain post for certification", zap.Error(err))
```

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,366 @@
+		c.client.RetryWaitMax = cfg.MaxRetryDelay
+	}
+}
+
+func NewCertifierClient(
+	logger *zap.Logger,
+	post *types.Post,
+	postInfo *types.PostInfo,
+	postCh []byte,
+	opts ...certifierClientOpts,
+) *CertifierClient {
```
We could use swagger/openapi to define the API for the certifier and have the client and server generated instead of implementing it by hand. Similarly how we use `buf` to generate the implementations for our `.proto` files:

Official go client generator: https://openapi-generator.tech/docs/generators/go There are at least one alternative as well: https://github.com/go-swagger/go-swagger

On the server side for rust I only found the official generator: https://openapi-generator.tech/docs/generators/rust-server/ I don't know if others exist.

poszu replied:
I could generate an OpenAPI doc from the rust code and include it in the release (there are instrumentation libraries to do this like https://lib.rs/crates/utoipa). Then, download the openapi doc in go-sm and build the Go client. I'm unsure it is worth the hassle given there is only a single endpoint.

fasmat replied:
I was rather thinking of generating both the client and server from an API definition and thereby already have documentation available (for possible alternative implementations).

We can do that at a different point in the future, I just thought having an API documentation as well would be nice 🙂

fasmat replied:
How to proceed with this? Having a spec and generate the code for client & server would also make the review easier/faster

poszu replied:
Having an OpenAPI spec would be nice, but it's not super important right now, so I'd rather not do it now.

fasmat replied:
Should we create an issue keeping track of this? I assume that issue would go into post-rs which would contain the OpenAPI spec from which the certifier server and the client in go-spacemesh are generated from?

poszu replied:
You can create an issue if you want.

fasmat replied:
created issue: https://github.com/spacemeshos/post-rs/issues/244

fasmat commented on activation/interface.go:
```diff
@@ -61,7 +62,7 @@ type layerClock interface {
 
 type nipostBuilder interface {
-	BuildNIPost(ctx context.Context, challenge *types.NIPostChallenge) (*types.NIPost, error)
+	BuildNIPost(ctx context.Context, challenge *types.NIPostChallenge, certifier certifierService) (*types.NIPost, error)
```
I feel like instantiating the certifier in the `atxBuilder` and passing it along to the nipostBuilder is more convoluted than it needs to be. Especially when we add multi-smesher support this will make the code harder to reason about and test.

What if the `CertifierService` would be its own service just like the `atxBuilder` that is instantiated during startup with the data it needs (e.g. PoET addresses and Smesher identities) and is constructor injected into the `nipostBuilder` so it can fetch the certificate from it when it submits to PoET?

fasmat replied:
This comment still applies, If we don't fix it we can add it as an issue to be refactored later.

poszu replied:
The problem with the approach you suggest is that the PoST, which is the crucial data for certifier is not available when the atx and nipost builders are constructed. That's why the certifier is created inside the atxBuilder, after the first PoST is ready.

fasmat replied:
Yeah but I thought that this will only be one of multiple possible ways to get a certificate to be used for a PoET?

Instead the certifier could just return a typed error `ErrMissingPost` or similar when the nipost builder requests a certificate for a poet and the certifier is unable to request that due to a missing post.

The initial post is always created and stored in the database before the certificate is requested anyway so theoretically this error should never even be needed to be returned. And the logic to fetch a post from an old ATX for the given identity should in my opinion also be part of the certifier rather than the nipost builder (and only after it doesn't find any post for the identity return an error).

fasmat commented on systest/Makefile:
```diff
@@ -5,7 +5,8 @@ tmpfile := $(shell mktemp /tmp/systest-XXX)
 image_name ?= $(org)/systest:$(version_info)
-poet_image ?= spacemeshos/poet:v0.9.7
+certifier_image ?= spacemeshos/certifier-service:bdbae9e09190a91d130fa6739493e391e0403a98
+poet_image ?= spacemeshos/poet:v0.10.0-rc2
```
Will there be a stable release of the two before the merge?

poszu replied:
Yes :)

poszu replied:
Done

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,366 @@
+	Pow     uint64 `json:"pow"`
+}
+
+type ProoToCertifyfMetadata struct {
```
```suggestion
type ProofToCertifyfMetadata struct {
```

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,366 @@
+	client certifierClient
+}
+
+type NewCertifierOption func(*Certifier)
```
```suggestion
type CertifierOption func(*Certifier)
```

poszu replied:
Why? This is an option to `NewCertifier()`, hence the name `NewCertifier+Option`.

fasmat replied:
It's configuring the Certifier not the `NewCertifier` function. `CertifierOption` would be in line with other functional option names we have used so far: either just `Opt`/`Option`/`OptionFunc` (if the package name identifies it uniquely) or `BuilderOption`, `PoetClientOpts`, `CertifierOpt`, `GeneratorOpt`, etc.

poszu replied:
IMHO, it's a "modifier" for the `NewCertifier()` function (changing its behavior). It's not a thing configuring a `Certifier` in a sense it's not meant to be called out of the `NewCertifier()` func like:

```go
c := NewCertifier()
WithCertificates(certs)(c)
```

Anyway, I will change it to be consistent with other Options in the code.

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,366 @@
+		poet   string
+	}
+
+	certifierInfos := iter.Map(poetsToCertify, func(p *PoetClient) *certInfo {
+		poet := *p
+		url, pubkey, err := poet.CertifierInfo(ctx)
+		if err != nil {
+			c.logger.Warn("failed to query for certifier info", zap.Error(err), zap.String("poet", poet.Address()))
+			return nil
+		}
+		return &certInfo{
+			url:    url,
+			pubkey: pubkey,
+			poet:   poet.Address(),
+		}
+	})
```
Why the extra dependency?

```suggestion
	certifierInfos := make([]*certInfo, len(poetsToCertify))
	var eg errgroup.Group
	for i := range poetsToCertify {
		i := i
		eg.Go(func() err {
			url, pubkey, err := poetsToCertify[i].CertifierInfo(ctx)
			if err != nil {
				c.logger.Warn("failed to query for certifier info", zap.Error(err), zap.String("poet", poetsToCertify[i].Address()))
				return nil
			}
			certifierInfos[i] = &certInfo{
				url:    url,
				pubkey: pubkey,
				poet:   poetsToCertify[i].Address(),
			}
		})
	}
	eg.Wait()
```

fasmat replied:
Alternatively I think it would also be good to have this function implemented once for one poet. Instead of iterating over all PoETs that went through a step (check if certificate is needed, fetch the CertifierInfo, fill the `certSvcs` map, Certify), it could instead call this one function for every PoET in an `errgroup` and return early when a PoET doesn't need to continue with the next step. Wdyt?

poszu replied:
I find the former version less cluttered, easier to read, and less error-prone.

>Alternatively I think it would also be good to have this function implemented once for one poet. Instead of iterating over all PoETs that went through a step (check if certificate is needed, fetch the CertifierInfo, fill the certSvcs map, Certify), it could instead call this one function for every PoET in an errgroup and return early when a PoET doesn't need to continue with the next step.

This code:
1. queries all poets for certifier info.
2. creates a unique set of certifiers to certify with
3. calls /certify on each

The point is to limit the number of calls on p. 3. All SM poets use the same certifier service and can reuse the same certificate so it makes sense to avoid duplicated calls to /certify.

fasmat replied:
> I find the former version less cluttered, easier to read, and less error-prone.

It's literally the same code 🤷  just instead of

```go
certifierInfos := iter.Map(poetsToCertify, func(p *PoetClient) *certInfo {
    ...
})
```
it's
```go
certifierInfos := make([]*certInfo, len(poetsToCertify))
var eg errgroup.Group
eg.Go(func() error {
    ... // (same as above)
})
eg.Wait()
```
But we avoid having to use an extra dependency?

poszu replied:
Then it's not literally the same code... (btw you missed the for loop).

What's wrong with a dependency? We depend on it indirectly already so using it doesn't add any build time. IMHO, the https://pkg.go.dev/github.com/sourcegraph/conc#section-readme library contains lots of useful building blocks making concurrent Go easier, less cluttered, and safer (given its 7.6k stars it's not something "random"). We could adapt it in the code base in more places.

fasmat replied:
I was referring to `...` as being the essentially the same code.

Generally I think we should aim to keep the amount of dependencies we use to a minimum. It might be the case that it is already a transient dependency, but those are (hopefully) managed by our direct dependencies. I don't think we need to start using a library for something that a) is simple to implement with the dependencies we already have and b) is arguably just as easy/hard to read:

The "self-made" approach is just using an `errgroup` in a for loop. In conctrast I had to check documentation / source code to see that `iter.Map` processes elements concurrently.

poszu replied:
>b) is arguably just as easy/hard to read

That's totally subjective. Also, one could say he needed to check `errgroup` documentation to see what it does (or any other library one is not familiar with yet) :shrug:. We already have 57 deps, some of which are doubtful (4 GH stars etc). In contrast, this is a well-known one with 7k+ stars.

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,366 @@
+}
+
+func (c *CertifierClient) Certify(ctx context.Context, url *url.URL, pubkey []byte) (PoetCert, error) {
+	request := CertifyRequest{
+		Proof: ProofToCertify{
+			Pow:     c.post.Pow,
+			Nonce:   c.post.Nonce,
+			Indices: c.post.Indices,
+		},
+		Metadata: ProoToCertifyfMetadata{
+			NodeId:          c.postInfo.NodeID[:],
+			CommitmentAtxId: c.postInfo.CommitmentATX[:],
+			NumUnits:        c.postInfo.NumUnits,
+			Challenge:       c.postCh,
+		},
+	}
```
I feel like these should be parameters to `Certify` instead of state of the client. Wdyt?

poszu replied:
I wanted to keep the `certifierClient` interface abstract of the details on _how_ the certificate is generated (basically to not include `Post` and `PostMetadata` in `Certify(...)`) to be able to support different methods of certification in the future. Hence `Certify()` only asks "certify with this certifier service" and the client needs to know how.

For now, the `Certifier` doesn't support many different clients to not complicate the code without a need.

What do you think? Perhaps you have a better idea for these interfaces?

fasmat reviewed (changes requested):
The changes in this PR will probably conflict with changes in https://github.com/spacemeshos/go-spacemesh/pull/5219/. We should align on how to resolve these :slightly_smiling_face:

I would prefer to have a DB migration once at startup that migrates a node to a state where the `atxBuilder`/`nipostBuilder`/`certifier` do not have to scan the DB looking for a PoST they can use to certify against a PoET, but rather set it up in a way where if a node has ever generated a PoST before the DB already contains that data. As a fallback we can fetch the PoST from the last published ATX by the node.

This will only cause issues in the case where we have a checkpoint and a user deleted their `node_state.sql` file right after the checkpoint. In this case they will probably have to generate an initial post again which is unfortunate but justifiable in my opinion.

fasmat commented on activation/e2e/certifier_client_test.go:
```diff
@@ -0,0 +1,188 @@
+	t.Cleanup(cleanup)
+	t.Cleanup(launchPostSupervisor(t, logger, mgr, grpcCfg, opts))
+
+	require.Eventually(t, func() bool {
+		_, err := svc.Client(sig.NodeID())
+		return err == nil
+	}, 10*time.Second, 100*time.Millisecond, "timed out waiting for connection")
+
+	postClient, err := svc.Client(sig.NodeID())
+	require.NoError(t, err)
```
```suggestion
	var postClient activation.PostClient
	require.Eventually(t, func() bool {
		var err error
		postClient, err = svc.Client(sig.NodeID())
		return err == nil
	}, 10*time.Second, 100*time.Millisecond, "timed out waiting for connection")
```

fasmat commented on activation/e2e/nipost_test.go:
```diff
@@ -177,161 +189,41 @@ func TestNIPostBuilderWithClients(t *testing.T) {
-		poetCfg,
-		mclock,
-	)
+	postClient, err := svc.Client(sig.NodeID())
```
Same as above

fasmat commented on activation/nipost.go:
```diff
@@ -291,6 +289,21 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.NIPos
 
 		nb.persistState()
+
+		post := nipost.Post{
+			Nonce:         proof.Nonce,
+			Indices:       proof.Indices,
+			Pow:           proof.Pow,
+			Challenge:     nb.state.PoetProofRef[:],
+			NumUnits:      postInfo.NumUnits,
+			CommitmentATX: postInfo.CommitmentATX,
+		}
+		if postInfo.Nonce != nil {
+			post.VRFNonce = *postInfo.Nonce
+		}
+		if err := nipost.AddPost(nb.localDB, nb.signer.NodeID(), post); err != nil {
+			nb.log.Warn("failed to persist post", zap.Error(err))
+		}
```
We should align on this, I don't think every post needs to be persisted on every `BuildNiPost`.

poszu replied:
We need to always have _some_ POST persisted to be able to recertify after a checkpoint recovery because the checkpoint recovery removes the NiPoST field from all existing ATXs.

fasmat replied:
See my other comments regarding `initial_post` and migrations. I just don't think we need to ever update the post that we persist for certificate purposes.

fasmat commented on activation/nipost.go:
```diff
@@ -310,50 +323,71 @@ func withConditionalTimeout(ctx context.Context, timeout time.Duration) (context
-	}
+	var round *types.PoetRound
 
+	for try := 0; try < 2; try++ {
+		round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nb.signer.NodeID(), auth)
+		if err == nil {
+			break
+		}
+		if errors.Is(err, ErrUnathorized) && try == 0 {
+			logger.Warn("failed to submit challenge as unathorized - recertifying", zap.Error(err))
+			auth.PoetCert, err = certifier.Recertify(ctx, client)
+			if err != nil {
+				return nil, &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
+			}
+		} else {
+			return nil, &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		}
+	}
```
This seems a bit too clever to me. Would it be possible to first check if we have a valid certificate, if not re-certify (return on an error) and then trying to submit (and return on an error)?

poszu replied:
Why is it "clever"? The code:
1. calls /submit with a certificate it already has in the DB
2. re-generates the certificate if the existing one turned out to be invalid

The code could be rewritten to avoid the loop :shrug:.

>Would it be possible to first check if we have a valid certificate, if not re-certify

I want to avoid an extra call to /info to obtain the certifier info before every /submit. In most cases, the certificate should be valid so the code optimistically assumes it and just re-certifies on an auth error.

fasmat replied:
I added a new comment for this - basically the code is a bit difficult to understand and I think it would be easier to put this logic into `client.Submit` (or some equivalent method on the certifier), i.e. fetching the certificate from DB and rectifying if needed

poszu replied:
The code is updated as you proposed in the other comment.

poszu committed:
Apply suggestions from code review

Co-authored-by: Matthias Fasching <5011972+fasmat@users.noreply.github.com>

poszu committed:
update nipost_test.go

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
fixup ProofToCertifyMetadata struct name

poszu committed:
Run poet systests in parallel

poszu committed:
Add NodeID to certs passed in config

poszu commented:
bors try

poszu committed:
Fix certifier config in systests

poszu commented:
bors try

poszu commented:
bors try

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,367 @@
+	return nil
+}
+
+type Certificate struct {
+	NodeID      types.NodeID `mapstructure:"node_id"`
+	Poet        string       `mapstructure:"poet"`
+	Certificate Base64Enc    `mapstructure:"certificate"`
+}
```
Is this part of the config? I feel like it shouldn't be:

```json
{
  "main": {
    ....
    "poet-server": [
      "https://mainnet-poet-0.spacemesh.network",
      "https://mainnet-poet-1.spacemesh.network",
      "https://mainnet-poet-2.spacemesh.network",
      "https://poet-110.spacemesh.network",
      "https://poet-111.spacemesh.network",
      "https://poet-112.spacemesh.network"
    ]
  },
  "poet": {
    "cycle-gap": "12h"
  },
  "certifier": {
    "certificates": [
      {
        "NodeID": "asdf1234",
        "Poet": "https://mainnet-poet-0.spacemesh.network",
        "Certificate": "aGVsbG8gd29ybGQ="
      }
      ...
    ]
  }
```

Every PoET is configured once as part of the node config and then again once for every node ID. Its a lot of redundant data and the config doesn't yet contain any node specific data  - in a sense one can copy it between nodes or share it.

poszu replied:
There is a unique certificate for each (nodeID, poet) pair - there's no redundant data and a certificate cannot be shared between nodes. 

The ability to use a certificate obtained externally is a requirement.

poszu replied:
Removed. It will be implemented in #5296

poszu committed:
Rename NewCertifierOption

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu commented:
bors try

poszu committed:
Apply localdb code migrations on checkpoint recovery

poszu commented:
bors try

poszu commented:
bors try

poszu commented:
bors try

poszu committed:
Remove importing certificates

poszu committed:
Build initial post if post cannot be found

poszu committed:
Remove sourcegraph/conc dep

poszu commented:
bors try

poszu committed:
Bump certifier-service in systests to v0.6.0

poszu commented:
bors try

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Don't update post in localdb

Persist POST extracted from existing ATX

poszu commented:
bors try

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Fix UT

poszu commented:
bors try

poszu committed:
Autoscale post verifying workers (#5354)

Verifying POST proofs is CPU-intensive. By default, it uses 3/4 CPU cores. It is a problem during the cycle gap for the nodes that take longer to generate the POST proof than the others - new ATXs start to drop in and verifying them fights for CPU with the proving process. In such a situation, proving is more important so sacrificing ATX verification speed is an acceptable tradeoff.

Added a way to scale the number of workers that verify POST proofs (in incoming ATXs) up and down.

There is an auto scaler that reacts to POST events (started/finished) by scaling the number of workers down/up respectively.

The minimum and maximum number of workers is configurable with `smeshing-opts-verifying-min-workers` and `smeshing-opts-verifying-workers` respectively.

Added unit tests.

- [ ] Update [changelog](../CHANGELOG.md) as needed

poszu commented:
bors test

poszu committed:
Fix default value of challenge in initial_post table

poszu commented:
bors try

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu commented:
bors try

poszu committed:
Fix flaky TestNIPostBuilder_Close

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Fixes

poszu committed:
Bump post and certifier services in systests

poszu commented:
bors try

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Support for poet certs with expiration

poszu commented:
bors try

poszu committed:
Use cert verification shared from poet

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Fix linting issues

poszu committed:
Bump certifier service in systests

poszu committed:
Rename local sql migration

poszu committed:
Fix staticcheck

poszu commented:
bors try

fasmat commented on systest/tests/poets_test.go:
```diff
@@ -203,3 +206,69 @@ func TestNodesUsingDifferentPoets(t *testing.T) {
+// Test verifying that nodes can register in both poets
+// - supporting PoW only
+// - supporting certificates
+// TODO: When PoW support is removed, convert this test to verify only the cert path.
```
Shall we add a GH issue for that and reference the issue ID in the todo?

poszu replied:
There is an issue for it, I can add a link in the comment. https://github.com/spacemeshos/go-spacemesh/issues/5212

fasmat commented on systest/Makefile:
```diff
@@ -5,7 +5,8 @@ tmpfile := $(shell mktemp /tmp/systest-XXX)
 image_name ?= $(org)/systest:$(version_info)
-poet_image ?= $(org)/poet:v0.10.2
+certifier_image ?= spacemeshos/certifier-service:51e174bbdb417bca21d4158ac04b48ebcb291fe1
+poet_image ?= $(org)/poet:v0.10.2-42-gceab169
```
This probably needs to be updated

fasmat commented on go.mod:
```diff
@@ -42,7 +42,7 @@ require (
 	github.com/spacemeshos/go-scale v1.1.13
 	github.com/spacemeshos/merkle-tree v0.2.3
-	github.com/spacemeshos/poet v0.10.2
+	github.com/spacemeshos/poet v0.10.3-0.20240325083037-7ba2761c0a35
```
Will PoET will be released first and then this updated?

poszu replied:
yes

fasmat commented on activation/activation.go:
```diff
@@ -346,35 +348,143 @@ func (b *Builder) buildInitialPost(ctx context.Context, nodeID types.NodeID) err
+	return atx.NIPost.Post, meta, atx.NIPost.PostMetadata.Challenge, nil
+}
+
+func (b *Builder) obtainPost(ctx context.Context, nodeID types.NodeID) (*types.Post, *types.PostInfo, []byte, error) {
```
Why not return a single object, we already have one that represents both `types.Post` and `types.PostInfo`:

```suggestion
func (b *Builder) obtainPost(ctx context.Context, nodeID types.NodeID) (*nipost.Post, error) {
```

fasmat commented on activation/activation.go:
```diff
@@ -346,35 +348,143 @@ func (b *Builder) buildInitialPost(ctx context.Context, nodeID types.NodeID) err
+func (b *Builder) obtainPostFromLastAtx(
+	ctx context.Context,
+	nodeId types.NodeID,
+) (*types.Post, *types.PostInfo, []byte, error) {
```
Same here, we already have a type that combines all the relevant information of a post in one type:

```suggestion
func (b *Builder) obtainPostFromLastAtx(ctx context.Context, nodeId types.NodeID) (*nipost.Post, error) {
```

fasmat commented on activation/activation.go:
```diff
@@ -346,35 +348,143 @@ func (b *Builder) buildInitialPost(ctx context.Context, nodeID types.NodeID) err
+	}
+
+	b.log.Info("POST not found in local DB. Trying to obtain POST from an existing ATX")
+	if post, postInfo, ch, err := b.obtainPostFromLastAtx(ctx, nodeID); err == nil {
```
`obtainPostFromLastAtx` could already return a `nipost.Post` and this function as well. Which would reduce the needed number of conversions from one type to the other.

fasmat commented on activation/activation.go:
```diff
@@ -346,35 +348,143 @@ func (b *Builder) buildInitialPost(ctx context.Context, nodeID types.NodeID) err
+// submitting to the poets.
+func (b *Builder) certifyPost(ctx context.Context, post *types.Post, meta *types.PostInfo, ch []byte) {
+	client := NewCertifierClient(b.log, post, meta, ch, WithCertifierClientConfig(b.certifierConfig.Client))
+	certifier := NewCertifier(b.localDB, b.log, client)
```
Why are those 2 types? I think outside of tests it makes no sense to have only a client or only a certifier.

If this is just for testing, you can still pass a mocked `client` in a test (either by overwriting the field or with a functional option).

fasmat commented on activation/activation_test.go:
```diff
@@ -881,10 +908,17 @@ func TestBuilder_PublishActivationTx_NoPrevATX_PublishFails_InitialPost_preserve
+		) (*nipost.NIPostState, error) {
+			return nil, ErrATXChallengeExpired
+		},
+	)
```
Why `DoAndReturn`?

```suggestion
	tab.mnipost.EXPECT().BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, ErrATXChallengeExpired)
```

fasmat commented on activation/activation_test.go:
```diff
@@ -1398,13 +1456,6 @@ func TestBuilder_RetryPublishActivationTx(t *testing.T) {
-	require.ErrorIs(t, err, sql.ErrNotFound)
-
-	_, err = nipost.Challenge(tab.localDB, sig.NodeID())
-	require.ErrorIs(t, err, sql.ErrNotFound)
```
Why? This should still be true, the ATX was published, the Challenge isn't needed any more (a new one will be built for the next ATX).

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+	}
+	e.Inner = b
+	return nil
+}
```
I don't see this type being used anywhere?

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+	}
+}
+
+type ProofToCertify struct {
+	Nonce   uint32 `json:"nonce"`
+	Indices []byte `json:"indices"`
+	Pow     uint64 `json:"pow"`
+}
+
+type ProofToCertifyMetadata struct {
+	NodeId          []byte `json:"node_id"`
+	CommitmentAtxId []byte `json:"commitment_atx_id"`
+
+	Challenge []byte `json:"challenge"`
+	NumUnits  uint32 `json:"num_units"`
+}
+
+type CertifyRequest struct {
+	Proof    ProofToCertify         `json:"proof"`
+	Metadata ProofToCertifyMetadata `json:"metadata"`
+}
```
This can be a single type that can be filled from the info that is in `nipost.Post`, the separation of Post into proof and metadata is arguably arbitrary and it would make (at least to me) a lot more sense if a PoST is self contained.

The separation was only because of the ATX wire type but that shouldn't use any of these types anyway 🙂 

```suggestion
type CertifyRequest struct {
	Nonce   uint32 `json:"nonce"`
	Indices []byte `json:"indices"`
	Pow     uint64 `json:"pow"`
	
	NodeId          []byte `json:"node_id"`
	CommitmentAtxId []byte `json:"commitment_atx_id"`

	Challenge []byte `json:"challenge"`
	NumUnits  uint32 `json:"num_units"`
}
```

poszu replied:
This structure is expected by the certifier service: https://github.com/spacemeshos/post-rs/blob/e42f504ed4b70ea6c5d28609d44bcd79096acf2c/certifier/src/certifier.rs#L17-L20

fasmat replied:
But that's an implementation detail that shouldn't influence API design?

poszu replied:
Wdym? These types are encoded as JSON and sent in the HTTP POST request.

fasmat replied:
I mean that the reason that `Post` and `PostMetadata` are separate objects is not because it makes sense from an (Certifier) API perspective but to safe on bytes in an ATX, where the metadata exists as a separate entity to deduplicate values that would be present twice in the ATX if they were part of the `Post` type again resulting from the fact that we used to re-use the same types for wire encoding and inside the domain of the codebase.

It's not a big issue if we at some point in the future re-do the API with OpenAPI anyway, I'm just arguing that if we look at it from a pure API perspective a PoST is a single object and data/metadata separation is arbitrary for the use in the Certifier.

EDIT: we can add this as a hint to the issue for using openapi in the future. 🙂

poszu replied:
It makes sense from the certifier API perspective because all the logic there also expects division into PoST (the proof) and metadata (the data required to verify it).

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+func NewCertifier(
+	db *localsql.Database,
+	logger *zap.Logger,
+	client certifierClient,
```
As mentioned earlier instead of passing a client to the certifier I think both should be the same object.

The client can still be mocked in tests by just overwriting the `client` field or passing the mock via functional option.

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+	return c
+}
+
+func (c *Certifier) GetCertificate(poet string) *certifier.PoetCert {
```
NIT:

```suggestion
func (c *Certifier) Certificate(poet string) *certifier.PoetCert {
```

I'm also not sure about the usefulness of the method in general. There are two places where it is called:

1. in the `NIPostBuilder` where it's only purpose is to swallow the `sql.ErrNoFound` error - arguably its simpler if the `NIPostBuilder` fetches the certificate directly from the DB and (debug) log if it (didn't) find it or if an error occurred
2. the certify method below - where it can just be inlined

poszu replied:
Its purpose is also to implement the `certifierService` interface. The fact that the cert is stored in the local DB is an implementation detail of the `Certifier` that implements this interface.

fasmat replied:
It seems convoluted from a users perspective. Ideally I would like to just pass whatever I want to submit to the poet client and the poet client handles the authentication.

Instead the nipost builder now has to 1. try to fetch a certificate and then pass that certificate to the poet client or 2. fail to fetch a certificate and then decide to generate one with `Recertify` or use PoW instead. Imo the poet client should do that for the user (either when instantiating it or when calling submit) - the nipost builder shouldn't care about communication details with poet.

poszu replied:
That would be best, but I wouldn't put it in the poet HTTP client, but create a _poet service_ thing containing all the poet-specific logic, including submitting to all known/registered Poets. The `nipostBuilder` would then only need to take the challenge and call `poetService.Submit(challenge)`. The logic inside `poetService` would figure out the details (PoW, Cert, registering in N poets, retries). But this is a bigger change and deserves a dedicated PR

fasmat replied:
Sure, if we do not include it here we should create an issue for it and tackle it in the future 🙂

poszu replied:
Created issue #5865.

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+	certifierInfos := make([]*certInfo, len(poetsToCertify))
+	var eg errgroup.Group
+	for i, poet := range poetsToCertify {
+		i, poet := i, poet
```
```suggestion
```

Not needed any more

fasmat commented on activation/certifier.go:
```diff
@@ -0,0 +1,369 @@
+		return nil, fmt.Errorf("request failed with code %d (message: %s)", resp.StatusCode, body)
+	}
+
+	certRespose := CertifyResponse{}
```
```suggestion
	certResponse := CertifyResponse{}
```

fasmat commented on activation/e2e/nipost_test.go:
```diff
@@ -176,173 +187,44 @@ func TestNIPostBuilderWithClients(t *testing.T) {
-	require.NoError(t, err)
-}
-
-func TestNIPostBuilder_Close(t *testing.T) {
```
Why was this test deleted? It verifies that creating a post can be canceled, something that wasn't possible at some time ago (or only with a big delay).

poszu replied:
It was not deleted but moved outside of end-to-end tests, where it doesn't belong.

fasmat commented on activation/e2e/nipost_test.go:
```diff
@@ -176,173 +187,44 @@ func TestNIPostBuilderWithClients(t *testing.T) {
-	require.Nil(t, nipost)
-}
-
-func TestNewNIPostBuilderNotInitialized(t *testing.T) {
```
This test also had a purpose. I'm not sure we still need it, but it tested that if PoST is not initialised while the `activation.Builder`/`nipostBuilder` were already querying for PoSTs it would not result in generating invalid posts but rather the builder waiting / retrying until PoST has finished initializing.

poszu replied:
This test didn't do what you described. There is a sequence:

```go
	t.Cleanup(launchPostSupervisor(t, logger, mgr, sig, grpcCfg, opts))

	require.Eventually(t, func() bool {
		_, err := svc.Client(sig.NodeID())
		return err == nil
	}, 10*time.Second, 100*time.Millisecond, "timed out waiting for connection")

	challenge := types.RandomHash()
	nipost, err := nb.BuildNIPost(context.Background(), sig, 7, challenge)
```

The `nb.BuildNIPost()` is called after initialization finished (after the post service already connected, and the service is only started when init finished).

I removed it because I found it unneccessary/duplicated.

poszu replied:
Even if I remove waiting for the post service to connect, is not guaranteed to test this. Most likely, the initialization will finish before poet and the `nipostBuilder` will try to build PoST when the init is already done.

fasmat replied:
No the waiting is required. The difference between this test and e.g. `TestNIPostBuilderWithClients` is that `TestNIPostBuilderWithClients` calls `initPost` before `launchPostSupervisor`; so the supervisor is started after the post data has already been completely generated.

This test did not do this, it just launched the Post Supervisor to ensure it will generate the data itself before actually starting the post service and waiting for the connection to the grpc service.

poszu replied:
I'm honestly confused. What is the purpose of the test? Is it to verify if `launchPostSupervisor` initializes the data on its own? Then it doesn't belong here and doesn't require a nipost builder at all.

fasmat replied:
Kind of. The `activation.PostSupervisor` is supposed to ensure that PoST is fully initialized before the post service is started, which is asserted by having the post service connect via grpc and then trying to generate a post.

I didn't see a test in `post_supervisor_test.go` or somewhere else where this is actually asserted besides this one test here. Alternatively you could remove the call to `initPost` from `TestNIPostBuilderWithClients` to not lose that implicit assertion/test and I create a small PR adding such a test again.

Wdyt?

fasmat replied:
Actually, nvm. I started writing a test for the post supervisor that would assert that data is initialized and noticed that tests for the `PostSetupManager` already do this on `StartSession`, while the `PostSupervisor` is tested to call `StartSession` before starting the post service process.

So no action necessary 🙂

fasmat commented on activation/interface.go:
```diff
@@ -117,27 +120,61 @@ type SmeshingProvider interface {
 }
 
-// poetClient servers as an interface to communicate with a PoET server.
+type PoetAuth struct {
+	*PoetPoW
+	*certifier.PoetCert
+}
```
This feels out of place here and probably belongs in `nipost.go`?

poszu replied:
It's placed here because it's part of the `PoetClient` interface below.

fasmat replied:
But shouldn't it go to the file where the client is implemented? `PoetPowParams` is also part of that interface and it its defined along the implementation.

poszu replied:
:shrug: 

I moved it to poet.go next to `PoetPow` and `PoetPowParams`.

fasmat commented on activation/interface.go:
```diff
@@ -117,27 +120,61 @@ type SmeshingProvider interface {
+// It holds the certificates and can recertify if needed.
+type certifierService interface {
+	// Acquire a certificate for the given poet.
+	GetCertificate(poet string) *certifier.PoetCert
```
NIT: avoid redundant prefixes for methods

```suggestion
	Certificate(poet string) *certifier.PoetCert
```

fasmat commented on activation/nipost.go:
```diff
@@ -346,50 +337,74 @@ func (nb *NIPostBuilder) submitPoetChallenge(
+		} else {
+			return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
+		}
+	}
```
I think we had this discussion before; the loop doesn't immediately make sense when reading the code. How about

```suggestion
	var (
		round *types.PoetRound
		err   error
	)
	round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, auth)
	switch {
	case errors.Is(err, ErrUnathorized):
		logger.Warn("failed to submit challenge as unathorized - recertifying", zap.Error(err))
		auth.PoetCert, err = certifier.Recertify(ctx, client)
		if err != nil {
			return &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
		}
		round, err = client.Submit(submitCtx, deadline, prefix, challenge, signature, nodeID, auth)
		if err != nil {
			return &PoetSvcUnstableError{msg: "failed to regenerate poet certificate", source: err}
		}
	case err != nil:
		return &PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
	default: // err == nil
	}
```

or even better, remove the logic for getting the certificate and adding it to `auth` from the builder and let the `client.Submit` method handle it by itself?

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

fasmat commented on activation/interface.go:
```diff
@@ -117,27 +120,61 @@ type SmeshingProvider interface {
+	Certify(ctx context.Context, url *url.URL, pubkey []byte) (*certifier.PoetCert, error)
+}
+
+// certifierService is used to certify nodeID for registerting in the poet.
```
```suggestion
// certifierService is used to certify nodeID for registering in the poet.
```

fasmat commented on activation/poet.go:
```diff
@@ -21,9 +21,8 @@ import (
-	ErrNotFound       = errors.New("not found")
-	ErrUnavailable    = errors.New("unavailable")
 	ErrInvalidRequest = errors.New("invalid request")
+	ErrUnathorized    = errors.New("unauthorized")
```
```suggestion
	ErrUnauthorized    = errors.New("unauthorized")
```

poszu committed:
Use nipost.Post

poszu committed:
More review fixes

poszu committed:
Update poet and post-rs

poszu commented:
bors try

fasmat commented on activation/activation_test.go:
```diff
@@ -1442,7 +1486,9 @@ func TestBuilder_InitialProofGeneratedOnce(t *testing.T) {
 	}
 	tab.mValidator.EXPECT().Post(gomock.Any(), sig.NodeID(), post.CommitmentATX, initialPost, meta, post.NumUnits).
 		Return(nil)
-	require.NoError(t, tab.buildInitialPost(context.Background(), sig.NodeID()))
+
+	_, _, err := tab.buildInitialPost(context.Background(), sig.NodeID())
+	require.NoError(t, err)
```
Shouldn't this be `obtainPost` rather than `buildInitialPost`?

poszu replied:
It doesn't matter really but I can change it.

fasmat commented on activation/poet.go:
```diff
@@ -151,27 +156,52 @@ func (c *HTTPPoetClient) PowParams(ctx context.Context) (*PoetPowParams, error)
-		Deadline: timestamppb.New(deadline),
+		Deadline:  timestamppb.New(deadline),
+	}
+	if auth.PoetPoW != nil {
+		request.PowParams = &rpcapi.PowParams{
+			Challenge:  auth.PoetPoW.Params.Challenge,
+			Difficulty: uint32(auth.PoetPoW.Params.Difficulty),
+		}
+		request.Nonce = auth.PoetPoW.Nonce
+	}
+	if auth.PoetCert != nil {
+		request.Certificate = &rpcapi.SubmitRequest_Certificate{
+			Data:      auth.PoetCert.Data,
+			Signature: auth.PoetCert.Signature,
+		}
 	}
```
What if both or neither are `nil`?

poszu replied:
Neither should happen because PoW is always and only ran if the cert is unavailable. However, assuming the caller is bad and it happens:
- both `nil` - the poet will reject registration,
- both are set - the poet will use the certificate.

Do you know a pattern/construct in Go to express a variant to enforce that always only one of them is set?

fasmat replied:
> Do you know a pattern/construct in Go to express a variant to enforce that always only one of them is set?

Using 2 methods instead of one 🙂

poszu committed:
update test

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Fix UTs

poszu committed:
Refactor certifier to lookup PoST on its own

poszu committed:
Fix systest

poszu commented:
bors try

fasmat commented on sql/atxs/atxs.go:
```diff
@@ -811,3 +811,17 @@ func PoetProofRef(ctx context.Context, db sql.Executor, id types.ATXID) (types.P
 	return types.PoetProofRef(atx.NIPost.PostMetadata.Challenge), nil
 }
+
+func Nipost(ctx context.Context, db sql.Executor, id types.ATXID) (*types.NIPost, error) {
```
The decoding of the nipost from the wire bytes shouldn't happen in the `sql/atxs` package. Instead the certifier should load the blob and decode the Nipost itself. `sql/atxs` should only be aware of domain and not wire types.

fasmat commented on activation/validation.go:
```diff
@@ -365,20 +364,6 @@ func (v *Validator) VerifyChain(ctx context.Context, id, goldenATXID types.ATXID
 	return v.verifyChainWithOpts(ctx, id, goldenATXID, options)
 }
 
-func (v *Validator) getNipost(ctx context.Context, id types.ATXID) (*types.NIPost, error) {
-	var blob sql.Blob
-	if err := atxs.LoadBlob(ctx, v.db, id.Bytes(), &blob); err != nil {
-		return nil, fmt.Errorf("getting blob for %s: %w", id, err)
-	}
-
-	// TODO: decide about version based on publish epoch
-	var atx wire.ActivationTxV1
-	if err := codec.Decode(blob.Bytes, &atx); err != nil {
-		return nil, fmt.Errorf("decoding ATX blob: %w", err)
-	}
-	return wire.NiPostFromWireV1(atx.NIPost), nil
-}
```
I'd prefer having it here or in the `activation.Certifier` instead of in the `sql/atxs` package.

poszu replied:
Moved to activation/certifier.go

poszu committed:
Merge remote-tracking branch 'origin/develop' into support-poet-certificates

poszu committed:
Review feedback

poszu committed:
Optimize searching for positioning ATX (#5952)

Searching for positioning ATX is slow because:
- the SQL query is slow
- it usually happens at the same time as many ATXs are being inserted into the DB (the poet CG)

poszu committed:
Move certification to the poet client

poszu committed:
Extend E2E activation test to use poet certificates with expiration

poszu committed:
Parallelize certifying initial post

poszu committed:
Avoid redundant certifications for (nodeID, pubkey) pairs

fasmat commented on activation/interface.go:
```diff
@@ -122,23 +122,18 @@ type SmeshingProvider interface {
 // It is used to submit challenges and fetch proofs.
 type PoetClient interface {
-	Address() string
+	Address() *url.URL
```
Why the change? Every call to `Address()` is followed by a call to `String()`.

poszu replied:
Not _every_. It seemed more correct to use url.URL type instead of string.

fasmat replied:
![Screenshot 2024-05-22 at 17 25 31](https://github.com/spacemeshos/go-spacemesh/assets/5011972/5f6666de-d99b-4e4e-8b9b-a7b091fb4394)

poszu replied:
Hmm, it was used as an URL but I must have removed that. Will revert to return string. Thanks

fasmat commented on activation/interface.go:
```diff
@@ -148,23 +143,27 @@ type PoetClient interface {
-	// It won't recertify poets that already have a certificate.
-	// It returns a map of a poet address to a certificate for it.
-	CertifyAll(ctx context.Context, id types.NodeID, poets []PoetClient) map[string]*certifier.PoetCert
+	certifierClient
```
I get the intention of using an embedded interface for simplicity but those are different concerns - the API of the (http) client isn't necessarily the same as for the service that provides the data to other parts of the node.

poszu replied:
Yeah, I wasn't sure about this. Currently, they are equal so I "inlined" it here. Should I define the `Certificate()` method separately in `certifierService`?

fasmat replied:
I think so... the two interfaces serve different purposes. They just happen to align in their method signature at the moment, but that could change in the future 🙂

fasmat commented on sql/localsql/certifier/db.go:
```diff
@@ -12,26 +12,26 @@ type PoetCert struct {
 }
 
-func AddCertificate(db sql.Executor, nodeID types.NodeID, cert PoetCert, poetUrl string) error {
+func AddCertificate(db sql.Executor, nodeID types.NodeID, cert PoetCert, cerifierID []byte) error {
```
We switched from poet ID to URL so we can more easily log and find the correct poet for the servers we submitted to, why change it back again?

poszu replied:
I changed here to certifier ID as a single certificate is valid for the pair (nodeID, certifier), not (nodeID, poet), because many poets can share the same certificate. Both certifier ID and URL could be used, I picked the ID as this will never change (the certificates would become invalid then). The URL, on the other hand, could change and the certificates would still be valid.

fasmat replied:
I'm not sure about this. Yes, certificates have an ID but they are usually bound to some domain. A certificate usually isn't valid for arbitrary domains even if the new domain points to the same service/key.

Also I am unsure about this feature - why would someone run a certifier service and then change its domain but nothing else about it?

poszu replied:
It's the certifier ID (its public key), not the certificate ID. The certificate is valid for a pair of (node ID and certifier pubkey), not (node ID and certifier domain). I'm not sure why would somebody change the domain (maybe forgot to pay and lost it ?), but that's fine and the certificates issued in the past **are still valid** from the poet's point of view. I see no reason to limit this on the node side.

poszu commented:
Bors try

fasmat commented on activation/certifier.go:
```diff
@@ -90,139 +93,67 @@ func NewCertifier(
+	certifier *url.URL,
+	pubkey []byte,
+) (*certifierdb.PoetCert, error) {
+	// We index certs in DB by node ID and pubkey. To avoid redundant queries, we allow only 1
+	// request per (nodeID, pubkey) pair to be in flight at a time.
+	key := string(append(id.Bytes(), pubkey...))
+	c.certificationsLock.Lock()
+	if ch, ok := c.certifications[key]; ok {
+		c.certificationsLock.Unlock()
+		<-ch
+	} else {
+		ch := make(chan struct{})
+		c.certifications[key] = ch
+		c.certificationsLock.Unlock()
+		defer func() {
+			c.certificationsLock.Lock()
+			close(ch)
+			delete(c.certifications, key)
+			c.certificationsLock.Unlock()
+		}()
+	}
```
Take a look at the `singleflight` package: https://pkg.go.dev/golang.org/x/sync/singleflight

Maybe it is worth considering using that over implementing something similar ourselves - we already have a dependency on the `golang.org/x/sync` package

poszu replied:
Changed to use this :+1:

fasmat commented on activation/poet.go:
```diff
@@ -338,25 +351,35 @@ func newPoetClient(db poetDbAPI, server types.PoetServer, cfg PoetConfig, logger
-		log.ZShortStringer("smesherID", nodeID),
-	)
+	challenge []byte,
+	logger *zap.Logger,
```
Why is this logger passed as argument? `poetClient` has a `logger` field?

poszu replied:
Because it gets a logger already enriched with additional fields like context and smesher ID.

fasmat replied:
Not part of this PR, but it doesn't make sense how we handle loggers and contexts to me at the moment. Ideally we would have a situation like this:

```go
ctx = log.WithFields(ctx, log.ZShortStringer("smesherID", nodeID) // probably already in node.go and passed to component
// ... some code to determine which poet to use
ctx = log.WithFields(ctx, zap.String("poet", client.Address())
// ... only pass ctx to child functions
// within child function
logger := c.log.WithOptions(log.ZContext(ctx))
logger.Info("some message") // prints message + smesherID field + poet address field
```

Right now we are passing both the logger and the context around and store information that should be added to logs in both the context and local loggers...

Wdyt?

poszu committed:
Review feedback

poszu committed:
Fix remaining UTs

poszu committed:
Fix remaining UTs

poszu commented:
bors merge

poszu commented:
bors merge
